packages/pigeon/lib/ast.dart

```diff
@@ -82,39 +74,84 @@ class Api extends Node {
   }
 }
 
-/// Represents a field on a [Class].
-class Field extends Node {
-  /// Parametric constructor for [Field].
-  Field({
-    required this.name,
+/// An entity that represents a typed concept, like a [TypeDeclaration] or [NamedType].
+abstract class TypedEntity {
+  /// The data-type of the entity (ex 'String' or 'int').
+  String get dataType;
+
+  /// The type arguments to the entity.
+  List<TypeDeclaration>? get typeArguments;
+
+  /// True if the type is nullable.
+  bool get isNullable;
+}
+
+/// A type.
+class TypeDeclaration implements TypedEntity {
+  /// Constructor for [TypeDeclaration].
+  TypeDeclaration({
     required this.dataType,
     required this.isNullable,
     this.typeArguments,
-    this.offset,
   });
 
+  /// A string representation of the base datatype.
+  @override
+  final String dataType;
+
+  /// The type arguments to this [TypeDeclaration].
+  @override
+  final List<TypeDeclaration>? typeArguments;
+
+  /// True if the type is nullable.
+  @override
+  final bool isNullable;
+
+  @override
+  String toString() {
+    return '(TypeDeclaration dataType:$dataType isNullable:$isNullable typeArguments:$typeArguments)';
+  }
+}
+
+/// Represents a type that has user defined name.
```

Comment:
It's vague on purpose.  My take away from our previous discussions was that you want to factor out similar datatypes, like TypeDeclaration.  The corollary to that is that we have datatypes that are versatile who don't represent the context they are used but the data they represent.  This could represent a field or an argument or a variable or a typedef.

Reply:
> It's vague on purpose.

It's not "vague", both the comment and the name are *actively misleading*. `String foo` is not "a named type", nor it is "a type that has user defined name". It is a variable, which has both a type and a name. The name in question is not a name *for* the type, which is what this comment says, but a name for the *instance*. The vague name for what you are describing would be something like `ThingWithTypeAndName`, which would (unlike `NamedType`) accurately describe both.

> My take away from our previous discussions was that you want to factor out similar datatypes, like TypeDeclaration. The corollary to that is that we have datatypes that are versatile who don't represent the context they are used but the data they represent.

That is not a corollary, it's a deliberate choice to go to an extreme (which I did not request) that has, in this context, hurt readability. The existence of TypeDeclaration does not mean that there can't be other types. You could absolutely have both a `Typedef` and a `Field` (or `Variable` or whatever) that happen to have the same two arguments for unrelated reasons. The fact that I want both to be implemented in terms of a `TypeDeclaration` rather than three identical interrelated variables that are duplicated in both places doesn't mean they can't exist. A typedef is conceptually very different from a field, local variable, or parameter, even if the base class fields happen to be the same.

Going back to my favorite trivial example: if code has a lot of x and y variable pairs and I say that really, `Point` is a foundational type in a codebase and should be created and used instead, it is not a corollary that well, actually, there are other things that have two related elements so rather than make `Point` the only conclusion is to make a generic `Pair` instead and have everything in the codebase that happens to have two related elements use that, such that it's not possible to easily distinguish between completely unrelated things that happen to have two elements.

Reply:
We talked offline and decided on an update on the docstring for NamedType, done.